### DISCUSSION TITLE
statuses/user_timeline too slow?

### DISCUSSION THREAD

**Discussion by Brando39** [tweepy/tweepy#1819] [Questions] [answered]
Hi!
I'm having some trouble lately getting the lastest tweet for some accounts. 
Basically I'm using the "statuses/user_timeline" API in "full speed" (at least until I finish the API calls) to monitor.
Usually I can get tweets very fast, but when an important and really desired tweet comes (for example a product get restocked and everybody wants it), the "statuses/user_timeline" API is very slow to get it (like half a minute).

This is a snipped of my code.
```
def tweetChecker(twitter_account):

    statuses = api.user_timeline(screen_name=twitter_account.account_name, count=1, tweet_mode="extended")
    tweet=statuses[0]
    if #tweet_is_new:
        return tweet


while(True):
    results=[]

    with concurrent.futures.ThreadPoolExecutor(max_workers=5)) as executor:
        results = executor.map(tweetChecker,twitter_account)

    for r in results:
        if r:
            #THERE'S A NEW TWEET!
```
My question are: 
Why those tweets comes so late? Is the API clogged by people accessing the same API with the same "screen_names" or am I using the twitter API wrong?

Keep in mind:
I don't use the streaming API because it feels slow in general
I'm using the auth-app API

Thanks to everyone and expecially to those who can help :)

**Comment by MichaelCurrin** [comment]
I haven't seen anyone else doing async stuff like this in tweepy outside of streaming. My guess is that Twitter is slowing you down from doing multiple requests at the same time even if you within the total limit.

Can I suggest looking up the profile for each user which you can do in a batch of 100 screen names at once. I may be mistaken but you find might the most recent tweet on the profile object. There is a tweet parameter on the method which also suggests this.

http://docs.tweepy.org/en/latest/api.html#API.lookup_users

**Comment by MichaelCurrin** [comment]
To be sure its not Twitter which is the bottleneck but your code implementation, can you run your tasks in parallel for some local long running tasks without the API? Such as printing and sleeping.

You can also have a look at what other solutions there are online for concurrent processing maybe with threading or multiprocessing libraries.

**Comment by Brando39** [comment]
> Can I suggest looking up the profile for each user which you can do in a batch of 100 screen names at once. I may be mistaken but you find might the most recent tweet on the profile object. There is a tweet parameter on the method which also suggests this.

You're right, I tested it and the User object return the last tweet/retweets. But why would this work better than user_timeline?

> To be sure its not Twitter which is the bottleneck but your code implementation, can you run your tasks in parallel for some local long running tasks without the API? Such as printing and sleeping.

I'm pretty sure my implementation in correct because:

- I already print in a .log file everytime the script checks on the profile. With that I can clearly see that it won't see the last tweet at the right time, but only seconds later

- With tweets which has no relevance it's always on point, but when an important ones comes, it doesn't see in time.

About the last point, my first thought was that the page was overloaded, could I be right?

Anyway thanks for taking the time to respond :)

**Comment by MichaelCurrin** [comment]
Thanks for confirming 

Getting a hundred user objects and their tweets is easy to do in a _single call_ with a list of 100 screen names. rather than doing 100 requests to user timeline in series or concurrently. So I suggested it as a very efficient solution that in guessing won't have your issue.

Even 500 users takes 5 requests. I'd still do in series first and then concurrently

Whatever the reason for the lack or performance on your side or Twitter limiting you I don't think you can get around it except by changing to the user lookup approach 

Twitter is able scale to millions of tweets made a day so my guess is that the bottleneck is on your side. I don't know how you chose 5 threads or what a good number is, but again run this against a non Twitter task to see how it performs at high volume. The problem is how to design a solution that mocks Twitter 

Twitter is probably able to service on account doing a bunch of requests at the same time , they might just limit that to prevent abuse.

**Comment by Brando39** [comment]
What you're suggesting is very clever!
I will definitely try this and I'll post an update as soon as I have some confirmation :)

**Comment by Brando39** [comment]
So today I've tested it and nothing changed. Normal tweets get instantly catched, but highly anticipated tweets get seen later.

```
while(True):
    results=[]

        users = api.lookup_users(screen_names=[account.account_name for account in twitter_account], tweet_mode="extended")

        results = map(#check if there's a new tweet for every account monitored)

        for r in results:
            if r:
                #Here's the new tweet
```

My code now looks like this, with no multithreading, too.
Is there an endpoint that is similiar to the push notification? Or a way to speed-up the streaming API? (unlikley, but asking costs nothing :P)

**Comment by MichaelCurrin** [comment]
Hmm so are you saying that when all of those accounts in your list in a short space of time, then status attribute is not set? And you get errors if using the old approach?

I don't know - maybe a twitter limatation

Using stream API to follow multiple accounts would work for your case, but I read it only gives 1-40% of tweets so is a lot less than the search API. The streaming API can't be sped up - but you can be efficient with your handling - using a queue manager to handle the tweets in a separate script using RabbitMQ, ZeroMQ etc. 
Or... you have to mess with multithreading or asyc or multiprocessing in a single script which is not nice. since streaming already has `is_async=True` feature so the process is non-blocking and that causes an error I've seen using Discord bot + Tweepy streaming together

Maybe that's good enough for you when you implement it.

**Comment by MichaelCurrin** [comment]
I'd avoid using a while loop and rather run the script at intervals e.g. every 5 minutes or 15 min or 1 hour.

Have a look at using list comprehension. You can do map and filter in one.


```
results = [user.status for user in users if user.status]
```

I don't know if user.status will be None, or give an attribute error so in the later case `getattr(user, "status", None)` which defaults to None if attribute is not set.

**Comment by Brando39** [comment]
> Hmm so are you saying that when all of those accounts in your list in a short space of time, then status attribute is not set? And you get errors if using the old approach?

I have trouble understanding your questions. The status attribute is always set, but appartently it doesn't change when an highly anticipated tweet arrives (a pratical example would be that the push notification, on this kind of tweets, arrives before my script, meanwhile on "normal" tweet my script is faster).

Catching some tweet faster and some slower makes me think it can't be twitter limiting me, neither my script. Begin these tweet very anticipated, maybe that account is overloaded with requests and the API struggle to return the newest tweet.

I already use the list comprehension on the function that map uses! I don't show that code beacuse it's just checking if the previous tweet id is the same as the current one

**Comment by Harmon758** [comment, marked as answer]
This is very likely a limitation of Twitter's API and not an issue with Tweepy.
You should ask about this on the [Twitter Developers Forums](https://twittercommunity.com/) instead.
